### ISSUE TITLE
Illegal instruction (core 0) at PC 0x00000000

### ISSUE THREAD

**Issue by Stronger-Huang** [QmppmQ/riscv#1]
make
./testbench
When I finished executing the above command, this problem occurred, 

```
Cycling clock to run for a few instructions
			     6000: Illegal instruction (core 0) at PC 0x00000000:
Halting
```

![VXTQKZEW_TPU1`_N19AY}}0](https://user-images.githubusercontent.com/61225949/165495227-d61c2ac3-0414-49c0-97ba-98584c60c7f6.png)
![(6WEXPZOE6RND @5ELS2LSY](https://user-images.githubusercontent.com/61225949/165495243-f75c1fed-e56a-4d23-9274-5d9d4d130b31.png)

It seems to be about instructions, but I haven't modified any files (including example.s and testbench).

I would appreciate it a lot if anyone can help me.
Thanks.

**Comment by QmppmQ** [comment]
The readme has not been updated for a while. I just updated the readme. You can try again with the new readme.
This is more complicated than before, but now it is possible to see the characters printed by the processor as it executes the program through the terminal, instead of having to verify the processor core or C code is working properly through the waveform diagram. Of course , if you want to see the waveform diagram, I will also update the steps in the readme later.

**Comment by Stronger-Huang** [comment]
Thank you! 
The program has problems: "BSP: illegal instruction exception handler entered", does it because the instruction added violates the basic RVFI instruction?

How can we verify the ten convolution outputs below and more importantly how to prove that the added instructions can indeed accelerate the CNN? 

![image](https://user-images.githubusercontent.com/61225949/167298592-8e863f52-c461-4e99-840d-61170f88bd42.png)

And can we see the waveform diagram and hardware resource overhead using Vivado?

Thanks again for your help!

**Comment by QmppmQ** [comment]
The added instruction does not violate the F extension, the opcode of added instruction is 0x77, and the  opcode of F extended is 0x53. You can check the opcode map in riscv-spec. 
![image](https://user-images.githubusercontent.com/56501722/167756368-f7b41a50-c30e-4813-a192-23c53cb13bf3.png)
**The reason for the illegal instruction is likely to be that the original rtl of cv32e40p has not been replaced with my modified one, and the original decoder cannot recognize the added instructions**.
The weights in the LeNet are trained with the fashion-mnist dataset, you can replace the weights or input in lenet.c with what you want. For reference, with the weights and input in the current code, the resulting output should be
![image](https://user-images.githubusercontent.com/56501722/167761217-9db157ab-a634-40f6-a359-60d9713ec114.png)
**We timing the classification task in lenet with traditional convolution and with custom convolution instructions, divide them to get the speedup.
The principle of custom instruction acceleration and winograd algorithm refers to this paper:**
Customized Instruction on RISC-V for Winograd-Based Convolution Acceleration
[https://ieeexplore.ieee.org/document/9516614](url)
This paper gives the speedup of the custom convolution instruction, as well as the waveform diagram, hardware resource overhead and power consumption using vivado.
![image](https://user-images.githubusercontent.com/56501722/167764127-3632a5ed-b081-43e7-9d30-e1f349ef4823.png)
![image](https://user-images.githubusercontent.com/56501722/167764144-f8216432-d7b3-4ece-aa85-40dd164375c7.png)

**This modified cv32e40p soft core also successfully runs on vivado, just convert the .hex or .elf file compiled by lenet.c into the corresponding .coe file and import it into the ram in the FPGA.**
If you use 100MHz clock to run it, it will takes more than 200ms to complete a classification task.

**Comment by Stronger-Huang** [comment]
Thank you so much, man. The opcode of 'reserved' can also be used for extension. It's working now. But could you please provide the source file of Vivado? I am not good at this software and it seems to have problems converting the .hex to .coe.

**Comment by QmppmQ** [comment]
I uploaded the files of the vivado project. If you directly use vivado to open the vivado project file *.xpr, you may encounter the following problems:
1. The version of vivado I used when I built the project is 2020.1. The later version of vivado should be able to be opened successfully, but the previous version may not be able to open it directly.
2. Most of the rtl files in the vivado project need to be imported again, the path is /FPGA/cv32e40p_rtl_bk_21.11.21.
3. I used the IP of block memory and clk, which is based on the pynq-z2 board, you need to regenerate the IP according to the FPGA board you are using

**Comment by Stronger-Huang** [comment]
Thank you, I upgraded two ips(blk_mem_gen_0 and clk_wiz_0). And load the lenet.coe through “Re-Customize IP ”blk_mem_gen_0“. 


![image](https://user-images.githubusercontent.com/61225949/170728854-b21d0290-b8f3-4aa1-8cfe-02af50de1f9b.png)


It comes with en error"Validation failed for parameter 'Coe File(Coe_File)' with value '../../../../cv32e40p_21.11.21.srcs/sources_1/lenet.coe' for IP 'blk_mem_gen_0'. The Memory Initialization vector can contain between 1 to Write Depth A number of entires. "

![image](https://user-images.githubusercontent.com/61225949/170728893-d19ad992-362f-4dc2-95fd-6fcf36ca6565.png)


Does it because the lines of lenet.coe(33907) be different from the write depth(16)? 
After I changed the "write depth" of Port A Options, the simulation results remain the same, and y4 to y0 are still "z". If I need to change something other about testbench?

![image](https://user-images.githubusercontent.com/61225949/170729767-e16f0d2f-b38e-4746-a5a4-235f3fe9a44b.png)

How can I solve it and simulate...

**Comment by QmppmQ** [comment]
You can refer to my setting of the memory
![1653706465(1)](https://user-images.githubusercontent.com/56501722/170807183-045cd262-36f7-465d-9869-d133b12091c8.jpg)
![1653706476(1)](https://user-images.githubusercontent.com/56501722/170807161-be192321-40b3-41d3-841d-93c2a0099836.jpg)
![1653706484(1)](https://user-images.githubusercontent.com/56501722/170807165-0bd31351-5f5e-4dc6-b298-cf4c71bc1ce0.jpg)
![1653706492(1)](https://user-images.githubusercontent.com/56501722/170807169-cbd710af-5a14-434b-baed-0d3068c93085.jpg)
![1653706500(1)](https://user-images.githubusercontent.com/56501722/170807172-94c81b20-3655-449c-ba1f-cd9cbff6a7d3.jpg)
CPU needs enough RAM space to run.
And if your clock frequency is 100MHz, the simulation takes more than 200ms to finish running to see the result. When you're running a behavioral simulation, timing violations don't need to be considered. But if you are running a post-synthesis timing simulation or post-implementation timing simulation, you need to make sure there are no timing violations first. You can avoid timing violations by reducing the core frequency.

**Comment by Stronger-Huang** [comment]
How can we compare the power consumption with and without the conv23 instruction? (to prove that with conv23 we can really save the power) 
Do we need to modify the lenet.c(replace the conv23 instruction with basic instruction) and convert it to .coe file and then load it in the ram?
By the way, what's the purpose of the file lenet_r.coe?


![image](https://user-images.githubusercontent.com/61225949/170850153-047430c5-9df4-46e1-8084-9b4b3115bedc.png)

**Comment by QmppmQ** [comment]
I guess what you want to know is how to compare the power consumption with and without the convolution module. Just instantiate convolution module mac_ops in ex_stage or not, and run implementation to see the power report in Vivado. Of course, the RI5CY with convolution module should consume more power than original core.
The answer of second question is yes, what a coincidence, that why lenet_r.coe is here.

**Comment by Stronger-Huang** [comment]
Thank you, I converted the input matrix in Lenet into an image, as shown below, which is a pair of sneakers, right? 

![colorimage](https://user-images.githubusercontent.com/61225949/170913752-d86bb4d0-bfd2-4055-9297-acb885a93ba8.jpg)

How can we test the output of lenet to get the accuracy of training data?

![image](https://user-images.githubusercontent.com/61225949/170914086-db94bef9-7226-42dd-af5a-f9025dc4d0f0.png)

**Comment by QmppmQ** [comment]
We train the Lenet by pytorch and put the trained weights into lenet.c. So accuracy is obtained by pytorch. The weights in lenet.c here have an accuracy of 87.8% for the Fusion-MNIST dataset.
